### PR TITLE
feat(hc): denormalize user and team into NotificationSetting

### PR DESCRIPTION
**Pull request by vbro**
This is the first step in breaking the foreign key relation between notification settings and team.
Design doc: https://www.notion.so/sentry/Notification-Settings-4f60428198db41a68ac887ac8a89f324

### CODE REVIEW

**Review by wedamija — COMMENTED** [review]
This table has around 2mm rows. Creating an index on these columns will likely take longer than our statement timeout (2s), and so the migration will fail.

I'd recommend you split out the index creation into a separate pr. One thing that is a little tricky here... If anyone deletes a ~team or~ user row, as part of that Django will follow the relation and check if any related rows exist in this table. Without the index we'll do a full table scan.

This might be ok, since deletes are relatively uncommon. But I'd recommend you set this up so that you merge/deploy this pr, and then ops creates the indexes on these columns as soon as the migration runs.

File: src/sentry/models/notificationsetting.py [hunk 1]
```diff
@@ -61,6 +62,8 @@ def provider_str(self) -> str:
     target = FlexibleForeignKey(
         "sentry.Actor", db_index=True, unique=False, null=False, on_delete=models.CASCADE
     )
+    team = FlexibleForeignKey("sentry.Team", null=True, db_index=False)
```
Comment (markstory) [inline src/sentry/models/notificationsetting.py:65]: I don't think this can be a foreign key as we'll need to break it again soon. Could we use HybridCloudForeignKey here instead.

File: src/sentry/models/team.py [hunk 2]
```diff
@@ -290,7 +292,8 @@ def transfer_to(self, organization):
             # we use a cursor here to avoid automatic cascading of relations
             # in Django
             try:
-                cursor.execute("DELETE FROM sentry_team WHERE id = %s", [self.id])
+                with in_test_psql_role_override("postgres"):
+                    cursor.execute("DELETE FROM sentry_team WHERE id = %s", [self.id])
```
Comment (markstory) [inline src/sentry/models/team.py:296]: Why is this necessary?

File: src/sentry/models/team.py [hunk 3]
```diff
@@ -315,9 +318,19 @@ def get_projects(self):
...
+            object_identifier=self.id,
+        )
+
     def delete(self, **kwargs):
         from sentry.models import ExternalActor
 
         # There is no foreign key relationship so we have to manually delete the ExternalActors
-        ExternalActor.objects.filter(actor_id=self.actor_id).delete()
-        return super().delete(**kwargs)
+        with transaction.atomic(), in_test_psql_role_override("postgres"):
```
Comment (markstory) [inline src/sentry/models/team.py:333]: Are we blocking delete on Team?

File: src/sentry/models/team.py (hunk 2 above)
Comment (vbro) [reply src/sentry/models/team.py]: This is to prevent `InsufficientPrivilege` in `test_team.TransferTest.test_existing_team`.
The `test_existing_team` test calls `team.transfer_to` which tries to delete a team via cursor.

File: src/sentry/models/team.py (hunk 3 above)
Comment (vbro) [reply src/sentry/models/team.py:336]: No. Delete on team should not be blocked. This new piece of code creates a RegionOutbox when deleting a Team object.

File: src/sentry/models/outbox.py [hunk 4]
```diff
@@ -58,6 +59,8 @@ class OutboxCategory(IntEnum):
     PROJECT_UPDATE = 8
     API_APPLICATION_UPDATE = 9
     SENTRY_APP_INSTALLATION_UPDATE = 10
+    TEAM_UPDATE = 11
+    TEAM_NO_CASCADE = 12
```
Comment (markstory) [inline src/sentry/models/outbox.py:63]: Do we need both of these? Having the cascade treated like other updates is preferable as it prevents us from processing a cascade *before* any update operations that may be waiting for delivery.

File: src/sentry/models/team.py [hunk 5]
```diff
@@ -290,7 +292,9 @@ def transfer_to(self, organization):
             # we use a cursor here to avoid automatic cascading of relations
             # in Django
             try:
-                cursor.execute("DELETE FROM sentry_team WHERE id = %s", [self.id])
+                with transaction.atomic(), in_test_psql_role_override("postgres"):
+                    cursor.execute("DELETE FROM sentry_team WHERE id = %s", [self.id])
+                    self.outbox_for_update(outbox_category=OutboxCategory.TEAM_NO_CASCADE).save()
```
Comment (markstory) [inline src/sentry/models/team.py:297]:
```suggestion
                    self.outbox_for_update(outbox_category=OutboxCategory.TEAM_UPDATE).save()
```

File: tests/sentry/models/test_notificationsetting.py [hunk 6]
```diff
@@ -53,3 +57,33 @@ def test_remove_for_organization(self):
         create_setting(actor=RpcActor.from_orm_user(self.user), organization=self.organization)
         self.organization.delete()
         assert_no_notification_settings()
+
+    def test_user_id(self):
```
Comment (mikejihbe) [inline tests/sentry/models/test_notificationsetting.py:61]: worth having a test for these deletion flows?

**Review by mikejihbe — COMMENTED** [review]
May need an outbox hander for team deletions as well?

**Review by mikejihbe — APPROVED** [review]
Just confirm that all tables that link to team.id as a FK are also bigints, then this looks fine to me

**Comment by wedamija** [comment]
> Just confirm that all tables that link to team.id as a FK are also bigints, then this looks fine to me

Just noticed the link to user is an `int`. I think our definition on `User` is wrong? In prod it's a bigint, but the field is just a bounded auto field. Not sure if that's a problem

File: tests/sentry/models/test_team.py [hunk 7]
```diff
@@ -150,3 +159,49 @@ def test_release_project_envs(self):
...
+
+        with outbox_runner():
+            team.delete()
+
+        assert not Team.objects.filter(id=team.id).exists()
+        assert NotificationSetting.objects.find_settings(
+            provider=ExternalProviders.EMAIL,
+            type=NotificationSettingTypes.ISSUE_ALERTS,
+            team=team,
+        ).exists()
```
Comment (vbro) [inline tests/sentry/models/test_team.py:197]: The test fails here because the NotificationSetting is cascade deleted with the team.
Any hints?
I commented out this line and verified that the call to `schedule_hybrid_cloud_foreign_key_jobs` below results in a call to `process_hybrid_cloud_foreign_key_cascade_batch` for `NotificationSetting.team_id`.

Comment (mikejihbe) [reply tests/sentry/models/test_team.py:197]: i'm confused, don't you _want_ this to be deleted too?

Comment (vbro) [reply tests/sentry/models/test_team.py:197]: Yes, it is deleted via Django cascade. I was wondering how to setup a test case to verify that the delete would be handled by `schedule_hybrid_cloud_foreign_key_jobs` in a control/region silo setup.
Perhaps this test is sufficient if I remove this line.

Comment (mikejihbe) [reply tests/sentry/models/test_team.py:197]: mark the test decorator with `stable=True` argument and the `outbox_runner` will take care of it.

Comment (vbro) [reply tests/sentry/models/test_team.py]: Does `stable=True` require a service for notification settings?
`notification_setting_service` is in progress. Merging this PR first.
